pycobertura/cobertura.py

```diff
@@ -41,37 +36,25 @@ class Cobertura(object):
     """
     An XML Cobertura parser.
     """
-    def __init__(self, report, source=None, source_prefix=None):
+    def __init__(self, report, filesystem=None):
         """
         Initialize a Cobertura report given a coverage report `report`. It can
         be either a file object or the path to an XML file that is in the
         Cobertura format.
 
-        The optional argument `source` is the location of the source code
-        provided as a directory path or a file object zip archive containing
-        the source code.
-
-        The optional argument `source_prefix` will be used to lookup source
-        files if a zip archive is provided and will be prepended to filenames
-        found in the coverage report.
+        The optional argument `filesystem` describes how to retrieve the source
+        files referenced in the report. Please check the pycobertura.filesystem
+        module in order to discover more about filesystems.
         """
         self.xml = ET.parse(report).getroot()
+        self._report = report if isinstance(report, basestring) else None
+        self._filesystem = filesystem
 
-        if source is None:
-            if isinstance(report, basestring):
-                # get the directory in which the coverage file lives
-                source = os.path.dirname(report)
-            self.filesystem = DirectoryFileSystem(
-                source, source_prefix=source_prefix
-            )
-        elif zipfile.is_zipfile(source):
-            self.filesystem = ZipFileSystem(
-                source, source_prefix=source_prefix
-            )
-        else:
-            self.filesystem = DirectoryFileSystem(
-                source, source_prefix=source_prefix
-            )
+    @property
+    def filesystem(self):
+        if not self._filesystem:
+            self._filesystem = filesystem_factory(self._report)
+        return self._filesystem
```

Comment:
What is the reason for adding this getter? Should we change the references in this file to access `self._filesystem` instead?

Reply:
Because when the Cobertura instance has been built with `Cobertura(xml)` there is no filesystem at all, and the two methods `file_source` and `source_lines` could throw. This property could be simply replaced with a 
`self.filesystem = filesystem if filesystem else filesystem_factory(report)` at line 51.
Or just dropped if you want your consumers to get exceptions when no filesystem has been passed at all.

Reply:
Gotcha.

Let's remove it as I would rather not expose this attribute as part of the public API. In hindsight, I should've made it internal/private in the first place. Also, now that the filesystem is being passed by the caller, they already have access to it should if they need the instance.

I would rather let it die or throw an exception that says that the Cobertura instance wasn't configured with a filesystem to be friendlier (but you don't have to do it, I could add that in a subsequent PR).